### PR TITLE
Handle closed, merged and reopened pull/merge requests in OBS workflows

### PR DESCRIPTION
**Pull request by dmarcoux**
Whenever the webhook event is for a closed, merged or reopened pull/merge requests, we handle this at the workflow level in the `Token::Workflow` model. There is no need to handle this in every step since we can already tell what has to be restored/deleted.

### CODE REVIEW

**Comment by dmarcoux** [comment]
[Support closing/merging/reopening actions for PRs/MRs](https://trello.com/c/nlIsH1Wp/1737-support-closing-merging-reopening-actions-for-prs-mrs)

File: src/api/app/models/token/workflow.rb [hunk 1]
```diff
@@ -13,13 +13,55 @@ def call(options)
...
+      Project.where(name: target_project_names).destroy_all
+    end
+  end
+
+  # TODO: This won't work for ConfigureRepositories since it doesn't have target_package and target_project_name
+  def restore_target_projects(workflows)
+    token_user_login = user.login
+
+    workflows.each do |workflow|
+      target_project_names = workflow.steps.map(&:target_project_name).uniq
```
Comment (hennevogel) [inline src/api/app/models/token/workflow.rb:47]: I think we should limit this to branch_package steps.

File: src/api/app/models/token/workflow.rb [hunk 2]
```diff
@@ -13,13 +13,55 @@ def call(options)
...
+  rescue Octokit::Unauthorized, Gitlab::Error::Unauthorized => e
+    raise Token::Errors::SCMTokenInvalid, e.message
+  end
+
+  private
+
+  # TODO: This won't work for ConfigureRepositories since it doesn't have target_package and target_project_name
+  def destroy_target_projects(workflows)
+    workflows.each do |workflow|
+      target_packages = workflow.steps.map(&:target_package).uniq
```
Comment (hennevogel) [inline src/api/app/models/token/workflow.rb:34]: I think we should limit this to branch_package steps.

File: src/api/app/models/token/workflow.rb (hunk 1 above)
Comment (dmarcoux) [reply src/api/app/models/token/workflow.rb]: We had a TODO about this and it's now done, we ignore `Workflow::Step::ConfigureRepositories` steps in this method. There's nothing to do for them.

File: src/api/app/models/token/workflow.rb (hunk 2 above)
Comment (dmarcoux) [reply src/api/app/models/token/workflow.rb]: We had a TODO about this and it's now done, we ignore `Workflow::Step::ConfigureRepositories` steps in this method. There's nothing to do for them.

File: src/api/app/models/workflow/step.rb [hunk 3]
```diff
@@ -16,6 +16,10 @@ def call(_options)
     raise AbstractMethodCalled
   end
 
+  def target_project_name
+    "home:#{@token.user.login}:#{source_project_name}:PR-#{scm_webhook.payload[:pr_number]}"
```
Comment (vpereira) [inline src/api/app/models/workflow/step.rb:20]: We have to access it from `Token::Workflow#restore_target_projects`

File: src/api/app/models/workflow/step.rb [hunk 4]
```diff
@@ -42,12 +42,12 @@ def target_package_name
     source_package_name
   end
 
-  private
-
   def target_package
     Package.find_by_project_and_name(target_project_name, target_package_name)
```
Comment (vpereira) [inline src/api/app/models/workflow/step.rb:46]: We have to access it from `Token::Workflow#destroy_target_projects`

File: src/api/app/models/workflow/step/link_package_step.rb [hunk 5]
```diff
@@ -7,8 +7,9 @@ def call(options = {})
 
     workflow_filters = options.fetch(:workflow_filters, {})
 
-    if scm_webhook.updated_pull_request? && target_package.present?
-      update_subscriptions(target_package, workflow_filters)
+    if scm_webhook.updated_pull_request?
+      create_target_package if target_package.blank?
+      create_or_update_subscriptions(target_package, workflow_filters)
```
Comment (saraycp) [inline src/api/app/models/workflow/step/link_package_step.rb:12]: According to this comment in [Workflow::Step model](https://github.com/openSUSE/open-build-service/blob/2acf0549256926c6b96190bab7a9f054e44ad447/src/api/app/models/workflow/step.rb#L87) there is a plan to remove the method `create_or_update_subscriptions` to use both `create_subscriptions` and `update_subscriptions` instead.

Comment (vpereira) [reply src/api/app/models/workflow/step/link_package_step.rb:12]: hm, could you explain why and who is it planned? Yesterday we were not sure why this `FIXME` was there.

Comment (saraycp) [reply src/api/app/models/workflow/step/link_package_step.rb:12]: When I worked on that code, I was suggested to not mix creating and updating to follow the [KISS principle](https://www.interaction-design.org/literature/topics/keep-it-simple-stupid).
As I was working on LinkPackageStep I postponed the refactoring of BrachPackageStep. I've decided to do it right now :).

Comment (saraycp) [reply src/api/app/models/workflow/step/link_package_step.rb:12]: In the new PR #11601, I refactor the BranchPackageStep the same way I did with LinkPackageStep in #11413

Feel free to base this PR on my changes or not, what is easier for you. I can always fix the conflicts in mine later.

Comment (dmarcoux) [reply src/api/app/models/workflow/step/link_package_step.rb:12]: We will tackle the refactoring of the BranchPackageStep after this PR is merged.

**Review by eduardoj — APPROVED** [review]

**Review by saraycp — APPROVED** [review]

**Comment by dmarcoux** [comment]
I reran test locally with some fake payload to mock closed/merged/reopened PRs and found a bug. I'll submit a fix.

>[8979d4b5-828d-4cd0-8316-b20061bd112d] [328:360.23] Completed 500 Internal Server Error in 12413ms (ActiveRecord: 26.2ms | Backend: 67.1ms | Allocations: 109806)
>
>
> [8979d4b5-828d-4cd0-8316-b20061bd112d] [328:360.23]   
> [8979d4b5-828d-4cd0-8316-b20061bd112d] NoMethodError (undefined method `create_or_update_subscriptions' for #<Token::Workflow:0x00007f94796c5510>
> Did you mean?  create_or_update):
> [8979d4b5-828d-4cd0-8316-b20061bd112d]   
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:66:in `block (2 levels) in restore_target_projects'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:65:in `each'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:65:in `block in restore_target_projects'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:56:in `each'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:56:in `restore_target_projects'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/token/workflow.rb:22:in `call'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/controllers/trigger_workflow_controller.rb:9:in `block in create'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/models/user.rb:865:in `run_as'
> [8979d4b5-828d-4cd0-8316-b20061bd112d] app/controllers/trigger_workflow_controller.rb:8:in `create'

**Comment by dmarcoux** [comment]
I changed `Token::Workflow#restore_target_projects` to call `create_or_update_subscriptions` with an instance from `Workflow::Step` since this method is from that model. `Workflow::Step#create_or_update_subscriptions` is now public, but this will have to be refactored.

**Review by eduardoj — APPROVED** [review]

**Pull request merged**